### PR TITLE
WaitForResourceHealthyAsync

### LINKED ISSUE #5469: ResourceNotificationService.WaitForResourceAsync is unreliable for Postgres resource
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Calling `await ResourceNotificationService.WaitForResourceAsync("postgres-db", KnownResourceStates.Running)` returns too early and results in cryptic exceptions when trying to actually connect to the database.

### Expected Behavior

The database should be fully accessible when this method has returned.

### Steps To Reproduce

This occurs when I am writing an integration test and need to do a database migration. Consider the following code:

**AppHost/Program.cs**
```csharp
var builder = DistributedApplication.CreateBuilder(args);
var pgsql = builder.AddPostgres("postgres");
var db = pgsql.AddDatabase("mydb");
var api = builder.AddProject<Projects.MyApi>("myapi").WithReference(db);
```

**MyTests.cs**
```csharp
public class MyTests
{
    [Fact]
    public async Task MyApiTest()
    {
        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AspireHost>();
        appHost.Services.ConfigureHttpClientDefaults(clientBuilder => { clientBuilder.AddStandardResilienceHandler(); });
        var app = await appHost.BuildAsync();
        await app.StartAsync();

        // waiting for all resources
        var rns = app.Services.GetRequiredService<ResourceNotificationService>();
        await rns.WaitForResourceAsync("mydb", KnownResourceStates.Running).WaitAsync(TimeSpan.FromSeconds(30));
        await rns.WaitForResourceAsync("myapi", KnownResourceStates.Running).WaitAsync(TimeSpan.FromSeconds(30));

        // ensure database is in actual state
        var cstr = await app.GetConnectionStringAsync("mydb")
        var opts = new DbContextOptionsBuilder<MyDbContext>().UseNpgsql(cstr).UseSnakeCaseNamingConvention().Options;
        using(var ctx = new MyDbContext(opts))
            await ctx.Database.MigrateAsync();      // <<< the exception happens here

        // test the API
        var httpClient = app.CreateHttpClient("myapi");
        // ...
    }
}
```

The `MyDbContext` is a EF Core `DbContext`.

An exception pops up when calling `ctx.Database.MigrateAsync()` (see below).

Key points are:
* The test can succeed randomly, maybe 1 or 2 times out of 10
* The test always succeeds if I put a `Task.Delay(1000)` before creating the context or wait on a breakpoint
* The issue does not reproduce in actual code, e.g. when the host is being run via F5 in VS and the API is set to migrate the database on startup

### Exceptions (if any)

```
Npgsql.NpgsqlException
Exception while reading from stream
   at Npgsql.Internal.NpgsqlReadBuffer.<Ensure>g__EnsureLong|55_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Npgsql.Internal.NpgsqlConnector.RawOpen(SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken, Boolean isFirstAttempt)
   at Npgsql.Internal.NpgsqlConnector.<Open>g__OpenCore|213_1(NpgsqlConnector conn, SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken, Boolean isFirstAttempt)
   at Npgsql.Internal.NpgsqlConnector.Open(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.UnpooledDataSource.Get(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlConnection.<Open>g__OpenAsync|42_0(Boolean async, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlDatabaseCreator.Exists(Boolean async, CancellationToken cancellationToken)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlDatabaseCreator.Exists(Boolean async, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.ExistsAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.GetAppliedMigrationsAsync(CancellationToken cancellationToken)
   at Npgsql.EntityFrameworkCore.PostgreSQL.Migrations.Internal.NpgsqlMigrator.MigrateAsync(String targetMigration, CancellationToken cancellationToken)

System.IO.EndOfStreamException
Attempted to read past the end of the stream.
   at Npgsql.Internal.NpgsqlReadBuffer.<Ensure>g__EnsureLong|55_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
```

### .NET Version info

```
.NET SDK:
 Version:           8.0.303
 Commit:            29ab8e3268
 Workload version:  8.0.300-manifests.56cd0383
 MSBuild version:   17.10.4+10fbfbf2e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.303\

.NET workloads installed:
 [aspire]
   Installation Source: SDK 8.0.300, VS 17.10.35122.118
   Manifest Version:    8.1.0/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.1.0\WorkloadManifest.json
   Install Type:        FileBased


Host:
  Version:      8.0.7
  Architecture: x64
  Commit:       2aade6beb0

.NET SDKs installed:
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.425 [C:\Program Files\dotnet\sdk]
  8.0.303 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

### Anything else?

Output for `dotnet workload list`:

```
Installed Workload Id      Manifest Version      Installation Source
--------------------------------------------------------------------------------
aspire                     8.1.0/8.0.100         SDK 8.0.300, VS 17.10.35122.118
```

The packages in the solution are v8.1.0.

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

Adds a `WaitForResourceHealthyAsync` method to `ResourceNotificationService`. Changed a number of tests to use it instead of the predicate version of `WaitForResourceAsync`.

Fixes #5469 

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No (existing test coverage switched to using it)
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No (this PR can serve as review)
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5867)

### CODE REVIEW

**Comment by davidfowl** [comment]
@radical lets start using this in more of our tests instead of waitfortext where possible.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 1]
```diff
@@ -168,6 +168,21 @@ static bool IsContinuableState(CustomResourceSnapshot snapshot) =>
...
+    /// Waits for a resource to become health.
+    /// </summary>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A task.</returns>
+    /// <remarks>
+    /// This method returns a task that will complete with the resource is healthy. A resource
+    /// without <see cref="HealthCheckAnnotation"/> annotations will be considered healthy.
+    /// </remarks>
+    public async Task WaitForResourceHealthyAsync(string resourceName, CancellationToken cancellationToken)
```
Comment (davidfowl) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:181]: I think these methods should return the snapshot.

**Comment by mitchdenny** [comment]
> @radical lets start using this in more of our tests instead of waitfortext where possible.

I changed a few tests over to use it where it was a quick change. In some cases like admin tools we don't have health checks so it wouldn't do what you want there (and we need to test both code paths anyway).

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 2]
```diff
@@ -168,6 +168,22 @@ static bool IsContinuableState(CustomResourceSnapshot snapshot) =>
             snapshot.State?.Text == KnownResourceStates.FailedToStart;
     }
 
+    /// <summary>
+    /// Waits for a resource to become health.
```
Comment (eerhardt) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:172]: Wording.
```suggestion
    /// Waits for a resource to become healthy.
```

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 3]
```diff
@@ -168,6 +168,22 @@ static bool IsContinuableState(CustomResourceSnapshot snapshot) =>
...
+    /// <returns>A task.</returns>
+    /// <remarks>
+    /// This method returns a task that will complete with the resource is healthy. A resource
+    /// without <see cref="HealthCheckAnnotation"/> annotations will be considered healthy.
+    /// </remarks>
+    public async Task<ResourceEvent> WaitForResourceHealthyAsync(string resourceName, CancellationToken cancellationToken)
+    {
+        var resourceEvent = await WaitForResourceAsync(resourceName, re => re.Snapshot.HealthStatus == HealthStatus.Healthy, cancellationToken: cancellationToken).ConfigureAwait(false);
+        return resourceEvent;
+    }
```
Comment (eerhardt) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:185]:
```suggestion
    public Task<ResourceEvent> WaitForResourceHealthyAsync(string resourceName, CancellationToken cancellationToken)
    {
        return WaitForResourceAsync(resourceName, re => re.Snapshot.HealthStatus == HealthStatus.Healthy, cancellationToken: cancellationToken);
    }
```

**Review by davidfowl — APPROVED** [review]

**Pull request merged**